opm/input/eclipse/Schedule/Action/ActionParser.cpp

```diff
@@ -17,324 +17,583 @@
-        return TokenType::op_eq;
+    auto tree = this->parse_or();
+
+    if (tree.type == Opm::Action::TokenType::error) {
+        throw std::invalid_argument {
+            "Failed to parse ACTIONX condition."
+        };
     }
 
-    if ((lower_arg == "!=") || (lower_arg == ".ne.")) {
-        return TokenType::op_ne;
+    if (this->currentType() != Opm::Action::TokenType::end) {
+        throw std::invalid_argument {
+            fmt::format("Extra unhandled data starting with "
+                        "token[{}] = {} in ACTIONX condition",
+                        this->current_pos_, this->currentToken())
+        };
     }
 
-    {
-        char* end_ptr = nullptr;
-        std::strtod(lower_arg.c_str(), &end_ptr);
+    return std::make_unique<Opm::Action::ASTNode>(std::move(tree));
+}
 
-        if (std::strlen(end_ptr) == 0) {
-            return TokenType::number;
-        }
+const std::string& ActionParser::currentToken() const
+{
+    if (this->parseIsComplete()) {
+        throw std::logic_error {
+            "Parse process proceeds beyond token sequence end"
+        };
     }
 
-    return TokenType::ecl_expr;
+    return this->tokens_[this->current_pos_];
 }
 
-FuncType Parser::get_func(const std::string& arg)
+Opm::Action::TokenType ActionParser::currentType() const
 {
-    if (arg == "YEAR") return FuncType::time;
-    if (arg == "MNTH") return FuncType::time_month;
-    if (arg == "DAY")  return FuncType::time;
-
-    using Cat = SummaryConfigNode::Category;
-    SummaryConfigNode::Category cat = parseKeywordCategory(arg);
-    switch (cat) {
-        case Cat::Aquifer:    return FuncType::aquifer;
-        case Cat::Well:       return FuncType::well;
-        case Cat::Group:      return FuncType::group;
-        case Cat::Connection: return FuncType::well_connection;
-        case Cat::Region:     return FuncType::region;
-        case Cat::Block:      return FuncType::block;
-        case Cat::Segment:    return FuncType::well_segment;
-        default:              return FuncType::none;
-    }
+    return this->parseIsComplete()
+        ? Opm::Action::TokenType::end
+        : Opm::Action::Parser::tokenType(this->currentToken());
 }
 
-ParseNode Parser::next()
+ActionParser::Node ActionParser::current() const
 {
-    this->current_pos++;
-    if (static_cast<size_t>(this->current_pos) == this->tokens.size()) {
-        return ParseNode(TokenType::end);
+    if (this->parseIsComplete()) {
+        return Node { Opm::Action::TokenType::end };
     }
 
-    std::string arg = this->tokens[this->current_pos];
-    return ParseNode(get_type(arg), arg);
+    const auto& arg = this->currentToken();
+    return Node { Opm::Action::Parser::tokenType(arg), arg };
 }
 
-ParseNode Parser::current() const
+bool ActionParser::parseIsComplete() const
 {
-    if (static_cast<size_t>(this->current_pos) == this->tokens.size()) {
-        return ParseNode(TokenType::end);
-    }
+    return static_cast<TokenSize>(this->current_pos_)
```

Comment:
not sure there is a problem but this can cast -1 to max(size_t) and then it would answer true.

Reply:
> not sure there is a problem but this can cast -1 to max(size_t) and then it would answer true.

I don't think that will happen in practice, because the first instruction of `buildParseTree()` is [a call to `next()`](https://github.com/OPM/opm-common/blob/bb8562aad5ff01521250380c8ecd432df3fe7519/opm/input/eclipse/Schedule/Action/ActionParser.cpp#L284) which advances the current position.  Nevertheless, I've `assert()`ed that the `current_pos_` is non-negative in my latest update to this PR.